app/services/kafka/policy_system_importer.rb

```diff
@@ -13,25 +13,35 @@ def initialize(message, logger)
     end
 
     def import
-      if V2::PolicySystem.exists?(policy_id: @policy_id, system_id: @system_id)
-        @logger.info("[#{@org_id}] PolicySystem for System #{@system_id} already exists")
-        return
-      end
+      return unless validate_sources!
 
-      ensure_exists(V2::Policy, @policy_id, 'Policy')
-      ensure_exists(V2::System, @system_id, 'System')
+      policy_system = V2::PolicySystem.new(policy_id: @policy_id, system_id: @system_id)
 
-      V2::PolicySystem.new(policy_id: @policy_id, system_id: @system_id).save!
-      @logger.audit_success("[#{@org_id}] Imported PolicySystem for System #{@system_id} from #{@msg_type} message")
+      if policy_system.save
+        @logger.audit_success("[#{@org_id}] Imported PolicySystem for System #{@system_id} from #{@msg_type} message")
+      else
+        audit_fail(policy_system.errors.full_messages.join(', ').to_s)
+      end
     end
 
     private
 
-    def ensure_exists(model, id, name)
-      return if model.exists?(id: id)
+    def validate_sources!
```

Comment:
```suggestion
    def validate_sources
```

The method is not modifying its object, which is what the exclamation mark represents

Reply:
It either returns boolean or raises, so I didn't quite agree with the question mark. 
I thought about the convention to mark "dangerous" methods with an exclamation mark, but perhaps here the raising of the exception only triggers retries, so it would be fine to have it with a question mark.